portal/static/js/admin.js

```diff
@@ -421,7 +421,91 @@ AdminTool.prototype.handleDownloadModal = function() {
     $("input[name='instrument'], input[name='downloadType']").on("click", function() {
         if ($(this).is(":checked")) $("#_downloadMessage").text("");
     });
-}
+};
+
+
+AdminTool.prototype.getDefaultTablePreference = function() {
+	return {sort_field: "id",sort_order: "desc"};
+};
+
+
+AdminTool.prototype.getTablePreference = function(userId, tableName) {
+    var prefData = null, self = this;
+    tnthAjax.getTablePreference(userId||self.userId, "patientList", {"sync": true}, function(data) {
+      if (data && !data.error) {
+        prefData = data || self.getDefaultTablePreference();
+        self.currentTablePreference = prefData;
+      };
+      //set filter values
+      self.setTableFilters(userId||self.userId);
+    });
+    return prefData;
+};
+
+AdminTool.prototype.setTableFilters = function(userId) {
+    var prefData = null;
+    if (this.currentTablePreference) {
+      prefData = this.currentTablePreference;
+    } else {
+      tnthAjax.getTablePreference(userId||this.userId, "patientList", {"sync": true}, function(data) {
+        if (data && !data.error) {
+          prefData = data;
+         };
+      });
+    };
+    if (prefData) {
+       //set filter values
+      if (prefData.filters) {
+        for (var item in prefData.filters) {
+          var fname = "#adminTable .bootstrap-table-filter-control-"+item;
+          /*
+           * note this is based on the trigger event for filtering specify in the plugin
+           */
+          $(fname).val(prefData.filters[item]).trigger($(fname).attr("type") == "text" ? "keyup": "change");
+        };
+      };
+    }
+};
+
+AdminTool.prototype.setTablePreference = function(userId, tableName, sortField, sortOrder) {
+  if (hasValue(tableName)) {
+    var data = {};
+    if (hasValue(sortField) && hasValue(sortOrder)) {
+      data["sort_field"] = sortField;
+      data["sort_order"] = sortOrder;
+    } else {
+    	//get selected sorted field information on UI
+    	var sortedField = $("#adminTable th[data-field]").has(".sortable.desc, .sortable.asc");
+    	if (sortedField.length > 0) {
+    		data["sort_field"] = sortedField.attr("data-field");
+    		var sortedOrder = "desc";
+    		sortedField.find(".sortable").each(function() {
+    			if ($(this).hasClass("desc")) sortedOrder = "desc";
+    			else if ($(this).hasClass("asc")) sortedOrder = "asc";
+    		});
+    		data["sort_order"] = sortedOrder;
```

Comment:
@OptimusRhine Could this cause a XSS if someone initially set their sorting preference for a column to be `alert("foo");`?

Reply:
Well, I don't think this is an issue with the sorting column/order, because on the backend, we only store sort_order values of 'asc' or 'desc' (anything else throws an error)

This COULD come up with the filter fields (basically, someone type `alert("foo")` into the filter field for a given column, and we'd display it later).

However, (a) I don't know if this is actually possible (depends on how those filter fields are called/rendered/etc), and (b) even if it is, it would only affect that user who inputted that value in the first place. I think XSS attacks are only considered dangerous, if a script one user puts in could influence what's displayed/rendered for another user (like by putting scripting into their username, which shows up in patient lists, etc). So even if it's possible here, I would think it's not a worry, as no user would be influenced besides the one generating the script in the first place...

Reply:
>only affect that user who inputted that value in the first place
Good point, had forgotten these would only be displayed for the associated user

>depends on how those filter fields are called/rendered/etc
Always keep in mind that bad guys can bypass the front-end validation entirely :)